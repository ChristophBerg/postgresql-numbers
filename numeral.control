default_version = '1'
comment = 'numeral datatypes extension'
relocatable = true
