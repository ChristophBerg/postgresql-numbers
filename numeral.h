/* type defs */

typedef long long Zahl;
typedef long long Roman;

/* prototypes */

const char *zahl_cstring (Zahl zahl);
int zahl_parse (char *s, Zahl *zahl);

const char *roman_cstring (Roman roman);
int roman_parse (char *s, Roman *roman);
