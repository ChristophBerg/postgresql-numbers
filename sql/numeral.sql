SELECT '0'::numeral;
SELECT '1'::numeral;
SELECT '-2'::numeral;

SELECT 'zero'::numeral;

SELECT 'one'::numeral;
SELECT 'minus one'::numeral;
SELECT 'two'::numeral;
SELECT 'three'::numeral;
SELECT 'four'::numeral;
SELECT 'five'::numeral;
SELECT 'six'::numeral;
SELECT 'seven'::numeral;
SELECT 'eight'::numeral;
SELECT 'nine'::numeral;
SELECT 'ten'::numeral;

SELECT 'eleven'::numeral;
SELECT 'twelve'::numeral;
SELECT 'thirteen'::numeral;
SELECT 'fourteen'::numeral;
SELECT 'fifteen'::numeral;
SELECT 'sixteen'::numeral;
SELECT 'seventeen'::numeral;
SELECT 'eighteen'::numeral;
SELECT 'nineteen'::numeral;

SELECT 'twenty'::numeral;
SELECT 'twenty-one'::numeral;
SELECT 'thirty'::numeral;
SELECT 'thirty-two'::numeral;
SELECT 'forty'::numeral;
SELECT 'fifty'::numeral;
SELECT 'sixty'::numeral;
SELECT 'seventy'::numeral;
SELECT 'eighty'::numeral;
SELECT 'ninety'::numeral;

SELECT 'hundred'::numeral;
SELECT 'one hundred'::numeral;
SELECT 'hundred one'::numeral;
SELECT 'one hundred one'::numeral;
SELECT 'one hundred twenty-one'::numeral;

SELECT 'two hundred'::numeral;
SELECT 'two hundred one'::numeral;
SELECT 'two hundred thirty-two'::numeral;

SELECT 'thousand'::numeral;
SELECT 'one thousand'::numeral;
SELECT 'one thousand one'::numeral;
SELECT 'thousand one hundred'::numeral;
SELECT 'one thousand one hundred'::numeral;
SELECT 'eleven hundred'::numeral;
SELECT 'one thousand one hundred one'::numeral;

SELECT 'hundred thousand'::numeral;
SELECT 'one hundred thousand'::numeral;
SELECT 'one hundred one thousand'::numeral;
SELECT 'two hundred thousand'::numeral;

SELECT 'one million'::numeral;
SELECT 'one million one'::numeral;
SELECT 'one million one hundred'::numeral;
SELECT 'one million one thousand'::numeral;
SELECT 'one million one hundred thousand'::numeral;
SELECT 'two million'::numeral;

SELECT 'hundred million'::numeral;

SELECT 'one billion'::numeral;
SELECT 'one billion one million'::numeral;
SELECT 'one billion one million one thousand'::numeral;

SELECT 'billion'::numeral;
SELECT 'trillion'::numeral;
SELECT 'quadrillion'::numeral;
SELECT 'quintillion'::numeral;
