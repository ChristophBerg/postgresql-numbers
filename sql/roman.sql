SELECT '0'::roman;
SELECT '1'::roman;
SELECT '-2'::roman;

SELECT 'nulla'::roman;
SELECT 'minus nulla'::roman;

SELECT 'I'::roman, 'i'::roman;
SELECT 'V'::roman, 'v'::roman;
SELECT 'X'::roman, 'x'::roman;
SELECT 'L'::roman, 'l'::roman;
SELECT 'C'::roman, 'c'::roman;
SELECT 'D'::roman, 'd'::roman;
SELECT 'M'::roman, 'm'::roman;

SELECT 'minus I'::roman;
SELECT 'minus 1'::roman;

SELECT 'II'::roman;
SELECT 'III'::roman;
SELECT 'IV'::roman;
SELECT 'V'::roman;
SELECT 'VI'::roman;
SELECT 'IX'::roman;
SELECT 'XI'::roman;
SELECT 'XX'::roman;
SELECT 'IL'::roman;
SELECT 'LI'::roman;
SELECT 'VL'::roman;
SELECT 'LV'::roman;
SELECT 'XL'::roman;
SELECT 'LX'::roman;
SELECT 'XC'::roman;
SELECT 'CX'::roman;
SELECT 'CC'::roman;
SELECT 'CD'::roman;
SELECT 'DC'::roman;
SELECT 'CM'::roman;
SELECT 'MC'::roman;
SELECT 'MM'::roman;

SELECT '10000'::roman;
SELECT '-10000'::roman;
SELECT '10001'::roman;
SELECT '-10001'::roman;
