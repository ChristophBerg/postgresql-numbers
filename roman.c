/*
Copyright (C) 2017 Christoph Berg

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
*/

#include "postgres.h"
#include "fmgr.h"
#include "numeral.h"

/* input and output */

char *yyerrstr; /* copy of error catched by yyromanerror() */

void yyromanerror (char *s);

void
yyromanerror (char *s)
{
	/* store error for later use in number_in */
	yyerrstr = pstrdup(s);
}

PG_FUNCTION_INFO_V1(roman_in);

Datum
roman_in (PG_FUNCTION_ARGS)
{
	char	*str = PG_GETARG_CSTRING(0);
	Roman	 roman;

	if (roman_parse(str, &roman) > 0)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
				 errmsg("invalid input syntax for type roman: \"%s\", %s",
					 str, yyerrstr)));

	PG_RETURN_INT64(roman);
}

PG_FUNCTION_INFO_V1(roman_out);

Datum
roman_out(PG_FUNCTION_ARGS)
{
	Roman	roman = PG_GETARG_INT64(0);
	PG_RETURN_CSTRING(roman_cstring(roman));
}

/* format roman numerals */

static char *
romanize (Roman roman)
{
	int val[] = { 1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1 };
	char *rom[] = { "M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I" };
	char res[1000] = "";
	int i;

	for (i = 0; i < 13; i++) {
		while (val[i] <= roman) {
			strlcat(res, rom[i], 1000);
			roman -= val[i];
		}
	}

	return pstrdup(res);
}

const char *
roman_cstring (Roman roman)
{
	if (roman < 0) {
		return psprintf("minus %s", roman_cstring(-roman));
	} else if (roman == 0) {
		return "nulla";
	}

	return romanize(roman);
}
