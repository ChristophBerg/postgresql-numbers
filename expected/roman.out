SELECT '0'::roman;
 roman 
-------
 nulla
(1 row)

SELECT '1'::roman;
 roman 
-------
 I
(1 row)

SELECT '-2'::roman;
  roman   
----------
 minus II
(1 row)

SELECT 'nulla'::roman;
 roman 
-------
 nulla
(1 row)

SELECT 'I'::roman, 'i'::roman;
 roman | roman 
-------+-------
 I     | I
(1 row)

SELECT 'V'::roman, 'v'::roman;
 roman | roman 
-------+-------
 V     | V
(1 row)

SELECT 'X'::roman, 'x'::roman;
 roman | roman 
-------+-------
 X     | X
(1 row)

SELECT 'L'::roman, 'l'::roman;
 roman | roman 
-------+-------
 L     | L
(1 row)

SELECT 'C'::roman, 'c'::roman;
 roman | roman 
-------+-------
 C     | C
(1 row)

SELECT 'D'::roman, 'd'::roman;
 roman | roman 
-------+-------
 D     | D
(1 row)

SELECT 'M'::roman, 'm'::roman;
 roman | roman 
-------+-------
 M     | M
(1 row)

SELECT 'minus I'::roman;
  roman  
---------
 minus I
(1 row)

SELECT 'II'::roman;
 roman 
-------
 II
(1 row)

SELECT 'III'::roman;
 roman 
-------
 III
(1 row)

SELECT 'IV'::roman;
 roman 
-------
 IV
(1 row)

SELECT 'V'::roman;
 roman 
-------
 V
(1 row)

SELECT 'VI'::roman;
 roman 
-------
 VI
(1 row)

SELECT 'IX'::roman;
 roman 
-------
 IX
(1 row)

SELECT 'XI'::roman;
 roman 
-------
 XI
(1 row)

SELECT 'XX'::roman;
 roman 
-------
 XX
(1 row)

SELECT 'IL'::roman;
 roman 
-------
 XLIX
(1 row)

SELECT 'LI'::roman;
 roman 
-------
 LI
(1 row)

SELECT 'VL'::roman;
 roman 
-------
 XLV
(1 row)

SELECT 'LV'::roman;
 roman 
-------
 LV
(1 row)

SELECT 'XL'::roman;
 roman 
-------
 XL
(1 row)

SELECT 'LX'::roman;
 roman 
-------
 LX
(1 row)

SELECT 'XC'::roman;
 roman 
-------
 XC
(1 row)

SELECT 'CX'::roman;
 roman 
-------
 CX
(1 row)

SELECT 'CC'::roman;
 roman 
-------
 CC
(1 row)

SELECT 'CD'::roman;
 roman 
-------
 CD
(1 row)

SELECT 'DC'::roman;
 roman 
-------
 DC
(1 row)

SELECT 'CM'::roman;
 roman 
-------
 CM
(1 row)

SELECT 'MC'::roman;
 roman 
-------
 MC
(1 row)

SELECT 'MM'::roman;
 roman 
-------
 MM
(1 row)

